keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,535 @@
+
+### Pod API
+
+`FileKeySelector` field will allow the users to select the value for environment
+variable from the specified key in the file. The user can specify a file using `VolumeName` and
+`Path` fields. The `VolumeName` field specifies the volume mount that contains
+the file and `Path` is the relative path in this volume mount filesystem.
+
+
+```
+type EnvVarSource struct {
+    ...
+    // Selects a key of the env file.
+    // +optional
+    FileKeyRef *FileKeySelector `json:"fileKeyRef,omitempty" protobuf:"bytes,5,opt,name=fileKeyRef"`
+    ...
+}
+
+type FileKeySelector struct {
+    // The name of the volume mount containing the env file.
+    VolumeName string `json:",inline" protobuf:"bytes,1,opt,name=volumeName"`
+    // The relative file path inside the volume mount to select from.
+    Path string `json:",inline" protobuf:"bytes,2,opt,name=path"`
+    // The key of the env file to select from.  Must be a valid key.
+    Key string `json:"key" protobuf:"bytes,3,opt,name=key"`
+    // Specify whether the file or its key must be defined. If the file or key
+    // does not exist, then the env var is not published.
+    // +optional
+    Optional *bool `json:"optional,omitempty" protobuf:"varint,4,opt,name=optional"`
+}
+```
+
+### Env File
+
+The full specification of an env file:
+
+1. **File Format**: The environment variable (env) file must adhere to valid [.env syntax](https://smartmob-rfc.readthedocs.io/en/latest/2-dotenv.html) to ensure correct parsing. An example:
+
+```
+KEY1=VALUE1
+KEY2=VALUE2
+...
+```
+
+2. **Variable Naming**: We will apply the same variable name [restrictions](https://github.com/kubernetes/kubernetes/blob/a7ca13ea29ba5b3c91fd293cdbaec8fb5b30cee2/pkg/apis/core/validation/validation.go#L2583-L2596) as other API-defined env vars.
+
+3. **Duplicate Names**: If an environment variable is defined multiple times in the file, the last occurrence takes precedence and overrides any previous values.
+
+4. **Size Limit**: To start with, the maximum allowed size for the env file will be 64KiB. Limits for key-value length will be added as a part of implementation after additional investigation.
+
+5. **File Location**: The env file must be placed within the `emptyDir` volume associated with the pod. If it is not found in the correct location, the Kubernetes API server will reject the pod creation request.
+
+6. **Container Behavior**:
+    
+    a. **Startup**: At container startup, the kubelet will parse the env file from the emptyDir volume and inject the defined variables into the container's environment. To avoid race condition with another container updating the env file, we will restrict mounting the emptyDir volume (containing the env file) in initContainer only. The env file can either be mounted or consumed in a container.
+    
+    b. **File Access**: The env file itself is not directly accessible within the container unless explicitly mounted by the container configuration.
```

Comment:
Is there any aspect of file permissions?  Does the file need to be readable by the pod's UID?  Or are we just ignoring that?

Reply:
If we respect file permissions on read, does this all get simpler?  Then we can just not care where the file came from at all - the pod is allowed to read it, so we read it.

It's one new failure mode in trade for getting rid of several.

Can we even find a way to test the permissions?

Reply:
kubelet is the one reading the env file and instantiate the pod container with env file. 

I believe kubelet which runs as root could possibly check that the pod container's UID is the owner of the env file before reading the file. WDYT?

Reply:
```$ stat file
....
Access: (0444/-r--r--r--)  Uid: (568056/testuser)   Gid: (89939/testuser)
....
```

kubelet could call stat on the env file and check its permission.
1. If the file is owned by pod UID or GID, then allow.
2. If the file can be read by everyone, then allow.
4. deny and fail container startup

Reply:
It's not clear to me if we actually have enough information to know the user's permissions nor that we have a mechanism to "sudo and try it".

My thinking was:
a) if the user is allowed to read the file then we can remove most or all other restrictions
b) if the user is NOT allowed to read the file, we should be REALLY careful about doing it for them (security)

The requirement that the file be in an emptyDir implies that the human creating the workload is allowed to do the thing they are trying to do, so I guess this is "safe enough", but it still makes me anxious.  @tallclair with your security hat on - tell me I am over-thinking it?

Reply:
@tallclair had a [comment](https://github.com/kubernetes/enhancements/pull/4524#pullrequestreview-2117049569) above.